scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,310 @@
+import heapq
+import itertools
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from scipy.integrate._rules import (
+    ProductNestedFixed, NestedFixedRule,
+    GaussKronrodQuad, NewtonCotesQuad,
+)
+
+
+@dataclass
+class CubatureRegion:
+    estimate: np.ndarray
+    error: np.ndarray
+    a: np.ndarray
+    b: np.ndarray
+
+    def __lt__(self, other):
+        # Consider regions with higher error estimates as being "less than" regions with
+        # lower order estimates, so that regions with high error estimates are placed at
+        # the top of the heap.
+        return _max_norm(self.error) > _max_norm(other.error)
+
+
+@dataclass
+class CubatureResult:
+    estimate: np.ndarray
+    error: np.ndarray
+    success: bool
+    status: str
+    regions: list[CubatureRegion]
+    subdivisions: int
+    atol: float
+    rtol: float
+
+
+def cub(f, a, b, rule="gk21", rtol=1e-05, atol=1e-08, max_subdivisions=10000,
+        args=(), kwargs=None):
```

Comment:
I guess my comment about h-adaptive vs p-adaptive was sparked by this interface. `max_subdivisions` is only relevant for h-adaptive methods. I wonder if the signature should only contain arguments that are relevant for all rules and then rule specific options should be passed in a dict. This is how the interface for `minimize` works for example.

Reply:
I think this is a good comment. I'm not going to suggest a solution, but here are ways it go. I think if we were to restrict to arguments valid to _all_ rules then we'd only have `f`, `a`, and `b`, not even `rtol` and `atol`. If we were to restrict to arguments valid to adaptive, either p-adaptive or h-adaptive, then we'd keep the tolerances as well. But then maybe in this case we'd just end up with two functions, `p_cubature` and`h_cubature`, with `h_cubature` being what is implemented here. Both could make use of the same rule interface.

Reply:
And you might still want `cubature` to alias to one of them, perhaps `h_cubature`.

Reply:
Agreed a few different ways to tackle this! Maybe wait and see what other people think